File: blind-75-Java/Monotonic/239.SlidingWindowMaximum.py
'''
239. Sliding Window Maximum
You are given an array of integers nums, there is a sliding window of size k which is moving from the very left of the array to the very right. You can only see the k numbers in the window. Each time the sliding window moves right by one position.

Return the max sliding window.

 

Example 1:

Input: nums = [1,3,-1,-3,5,3,6,7], k = 3
Output: [3,3,5,5,6,7]
Explanation: 
Window position                Max
---------------               -----
[1  3  -1] -3  5  3  6  7       3
 1 [3  -1  -3] 5  3  6  7       3
 1  3 [-1  -3  5] 3  6  7       5
 1  3  -1 [-3  5  3] 6  7       5
 1  3  -1  -3 [5  3  6] 7       6
 1  3  -1  -3  5 [3  6  7]      7
Example 2:

Input: nums = [1], k = 1
Output: [1]
 

Constraints:

1 <= nums.length <= 105
-104 <= nums[i] <= 104
1 <= k <= nums.length

'''
'''
The time complexity of the deque-based maxSlidingWindow 
algorithm is O(n), where n is the length of the input array. 
This efficiency comes from the fact that each element is
added to and removed from the deque at most once during
the entire process, resulting in a total of at most 2n operations.

Each index is pushed and popped from the deque at most once.

All other operations inside the loop (comparisons, 
appends to result) are O(1).

The space complexity is O(k), where k is the window size. 
This is because the deque stores at most k indices at 
any time (the indices of elements in the current window).
The output list also takes O(n - k + 1) space, but if only 
the auxiliary space is considered (not the output), it is O(k).

Summary Table:

Complexity	           Value	        Why
Time Complexity	        O(n)	Each element added/removed at most once
Space Complexity	    O(k)	Deque holds at most k indices

'''
from collections import deque
from typing import List

class Solution:
    def maxSlidingWindow(self,nums:List[int],k:int) -> List[int]:
        n = len(nums)
        if n == 0:
            return []
        deq= deque()
        result = []
        for i in range(n):
        # Remove indices from the front that are out of the current window
            while deq and deq[0] <= i-k:
                deq.popleft()
        # Remove indices from the back while the current element is greater
            while deq and nums[i] > nums[deq[-1]]:
                deq.pop()
            deq.append(i)
        # Append the current max to the result once the first window is complete
            if i>= k-1:
                result.append(nums[deq[0]])
        return result


nums = [1,3,-1,-3,5,3,6,7]
k=3
sol = Solution()
print(sol.maxSlidingWindow(nums,k))

File: blind-75-Java/Monotonic/239.SlidingWindowMaximum.js
/*
239. Sliding Window Maximum

You are given an array of integers nums, 
there is a sliding window of size k which is
 moving from the very left of the array to 
 the very right. You can only see the k
  numbers in the window. Each time the 
  sliding window moves right by one position.

Return the max sliding window.

Example 1:

Input: nums = [1,3,-1,-3,5,3,6,7], k = 3
Output: [3,3,5,5,6,7]
Explanation: 
Window position                Max
---------------               -----
[1  3  -1] -3  5  3  6  7       3
 1 [3  -1  -3] 5  3  6  7       3
 1  3 [-1  -3  5] 3  6  7       5
 1  3  -1 [-3  5  3] 6  7       5
 1  3  -1  -3 [5  3  6] 7       6
 1  3  -1  -3  5 [3  6  7]      7
Example 2:

Input: nums = [1], k = 1
Output: [1]
 

Constraints:

1 <= nums.length <= 105
-104 <= nums[i] <= 104
1 <= k <= nums.length

*/

var maxSlidingWindow = function (nums, k) {
    let n = nums.length;
    //Edge case: if the array is empty,
    //  return an empty array (no windows are possible).
    if (n === 0) return [];
    //Initialize an empty deque (array used as a queue).

    //Important: it stores indices of elements,
    //  not the actual values.
    //Storing indices helps us quickly check if an 
    // element is outside the current window.
    let deq = [];   // will store indices
    //Initialize an empty array result to
    // store the maximum of each sliding window.
    let result = [];

    for (let i = 0; i < n; i++) {
        // remove elements out of the current window
        //deq is the index of the oldest element in the deque.
//If this index is outside the current window 
// (<= i - k), remove it (shift() removes from the front).
//This ensures that all indices in deq are 
// within the current window of size k.
        while (deq.length > 0 && deq[0] <= i - k) {
            deq.shift();
        }
//While the deque is not empty and the current element 
// nums[i] is greater than the element corresponding
//  to the index at the back of deque, pop from the back.

//Why? Because if nums[i] is bigger, then 
// the older smaller elements cannot ever be
//  the maximum in this or any future 
// window → they are useless, so we discard them.
        // maintain deque in descending order (by values in nums)
        //deq.length - 1 gets the last index in the deque.
// deq[deq.length - 1] gives you the last value stored in the deque, which is an index in nums.
// nums[deq[deq.length - 1]] accesses the actual value in the nums array at the position pointed to by that index.

        while (deq.length > 0 && nums[i] > nums[deq[deq.length - 1]]) {
          
            deq.pop();
        }
//Add the current index i to the back of the deque.
//Now deq contains candidate indices for the maximum,
//  in descending order of their corresponding nums values.
        deq.push(i);
        //Once we've processed at least k elements
        //  (i.e., formed a full window), we can record a maximum.
        // add result when window is at least size k
        //deq is the index of the maximum element for the current window, because:

        // The deque is kept in decreasing order,

        // The front always represents the largest
        //  element that’s still inside the window.

        // Push that maximum (nums[deq]) into the result array.
        if (i >= k - 1) {
            result.push(nums[deq[0]]);
        }
    }

    return result;
}

let nums = [1,3,-1,-3,5,3,6,7], k = 3;
let res = maxSlidingWindow(nums,k);

console.log("maxSlidingWindow==",res)